### DISCUSSION TITLE
Remix - purely server side (html over wire)

### DISCUSSION THREAD

**Discussion by ajaishankar** [remix-run/remix#7422] [General]
Hi All

Huge remix fan here!

By having a clear separation of where data access happens (loaders and actions) it has made my life infinitely simpler, thanks!

Just wanted to check if anyone has experimented with using Remix purely on server side to emit HTML.

That is, avoid client side Remix - no React, hydration or anything...

1. Remix sends HTML over the wire
2. Client is Turbo (https://turbo.hotwired.dev) + Stimulus

I know it sounds like fighting the framework, but just asking... 😄 

Thanks

Ajai

**Comment by kiliman** [comment]
Just omit the `<Scripts/>` component in your *root.tsx* route. Remix will be entirely server-side.

All links and forms will continue to function as they are native HTML elements.

**Reply by ajaishankar** [reply]
@kiliman just wanted to follow up

Is it possible to configure Remix to build some extra files?

Say I omit the `<Scripts/>` but still need the Stimulus controllers (in Typescript) to be transpiled, minified and bundled?

```
app.ts
controllers
  cart-controller.ts
  cart-item-controller.ts
```

I'd like app.ts to be available as

```
<script type="module" src="app-ABCDEFGH.js">
```

**Reply by kiliman** [reply]
You may have better luck trying the new Vite support. 

https://github.com/remix-run/remix/discussions/2427

**Reply by ajaishankar** [reply]
Thanks @kiliman - definitely looks like I'm trying to abuse Remix 😄

**Reply by ajaishankar** [reply]
@kiliman now that [Vite support is stable](https://remix.run/blog/remix-vite-stable) wanted to revisit this.

> Just omit the `<Scripts/>` component in your _root.tsx_ route. Remix will be entirely server-side.

The `<Scripts>` component I believe is necessary for the [Vite integration](https://remix.run/docs/en/main/guides/vite#hmr--hdr)

So wanted to check if is there is still a simple way to omit just Remix client side?

**Reply by kiliman** [reply]
Hmm... not sure. It's not something I've tried before.

You could just render `<Scripts>` in development. For production, it would be entirely server-rendered.

```ts
{process.env.NODE_ENV === 'development' && <Scripts />}
```

**Reply by ajaishankar** [reply]
Thanks! will give this a try...

**Comment by ajaishankar** [comment]
wow - that simple! Thanks @kiliman

**Reply by frontsideair** [reply]
This would be an interesting mid-point by the way: https://twitter.com/ryanflorence/status/1703970940030496934

**Comment by ajaishankar** [comment]
One more question - how would I go about rendering a partial (without the layout)?

**Reply by sergiodxa** [reply]
You can use a resource route to send partials of HTML.

**Reply by ajaishankar** [reply]
I'd still like to use react components to render the html server side.

Is this possible in resource routes?

Or can we have a layoutless UI route where even the `root.tsx` is not rendered?

**Reply by sergiodxa** [reply]
Yes, you can do something like this
```ts
// app/routes/partial.something.tsx
import { renderToString } from "react-dom/server"
import { html } from "remix-utils"

import { Something } from "~/partial/something"

export async function loader() {
  return html(renderToString(<Something />))
}
```

**Reply by ajaishankar** [reply]
Thanks so much @sergiodxa!

**Comment by ajaishankar** [comment]
@kiliman @sergiodxa 

I was able to get this working by using the manifest from `useRemixContext` (2 line change)!

For local dev am able to see changes reflected as soon as a file is changed.

Also once built the html correctly references the javascript and css assets.

**entry.client.tsx**

Just has the client side code I need (hotwire turbo)

```ts
import * as Turbo_ from "@hotwired/turbo"

export const Turbo = Turbo_
```

**root.tsx**

Do not render `<Scripts />` tag

Get the manifest from `useRemixContext` to render the client entry module.

```ts
import { useRemixContext } from "@remix-run/react/dist/components";

export function Layout({ children }: { children: React.ReactNode }) {
  const { manifest } = useRemixContext();

  return (
    <html lang="en">
      <body>
        {children}
        <ScrollRestoration />
         {import.meta.env.DEV && <script type="module" src="/@vite/client"></script>}
        <script type="module" src={manifest.entry.module}></script>
      </body>
    </html>
  );
}
```

Once built the html correctly references the build entry client and also other assets like tailwind

```
<script type="module" src="/assets/entry.client-BI-UUuYH.js"></script>
<link rel="stylesheet" href="/assets/tailwind-DIvsNEUZ.css"/>
```

I _think_ this is a working solution, do suggest if I'm missing anything obvious...
